# LP Practice

In [39]:
import json

## What is the top level of the JSON data??

In [40]:
# Open the JSON file and display the first level of keys
# Open API results WITH json module
with open('data/example_yelp_results.json') as f:   #adjust for your path
    json_file = json.load(f)
    
## What type is top-level of json?
type(json_file)

dict

In [41]:
# what are the keys?
json_file.keys()

dict_keys(['businesses', 'total', 'region'])

## What is stored in the second level??
Three keys to explore... 

### Region

In [42]:
# What type is region?
type(json_file['region'])

dict

In [43]:
# what keys are in region?
json_file['region'].keys()

dict_keys(['center'])

In [44]:
# what type is center?
type(json_file['region']['center'])

dict

In [45]:
# What keys are in center?
json_file['region']['center'].keys()

dict_keys(['longitude', 'latitude'])

In [46]:
# visualize the whole center dict...
json_file['region']['center']

{'longitude': -76.607666015625, 'latitude': 39.29617825404647}

looks like the center point of the yelp search was lat -76, long 39

### total

In [47]:
# What is stored in the total key
type(json_file['total'])

int

In [48]:
# What is the value?
json_file['total']

435

looks like there were 435 results for the yelp search

### businesses

In [49]:
# What is in the businesses key?
type(json_file['businesses'])

list

In [50]:
# how long is the list of businesses
len(json_file['businesses'])

20

looks like there are 20 businesses stored in the list

In [51]:
# What does the first entry of the list look like??
json_file['businesses'][0]

{'id': 'D9A33FM394q99o4QtK5YwA',
 'alias': 'faidleys-seafood-baltimore-3',
 'name': 'Faidleys Seafood',
 'image_url': 'https://s3-media3.fl.yelpcdn.com/bphoto/OTjVDCVS7pGopH6GZcfjBA/o.jpg',
 'is_closed': False,
 'url': 'https://www.yelp.com/biz/faidleys-seafood-baltimore-3?adjust_creative=KJtcufKUS887p24u6rjVIQ&utm_campaign=yelp_api_v3&utm_medium=api_v3_business_search&utm_source=KJtcufKUS887p24u6rjVIQ',
 'review_count': 1184,
 'categories': [{'alias': 'seafood', 'title': 'Seafood'},
  {'alias': 'beerbar', 'title': 'Beer Bar'}],
 'rating': 4.0,
 'coordinates': {'latitude': 39.291696, 'longitude': -76.62224},
 'transactions': ['delivery'],
 'price': '$$',
 'location': {'address1': '203 N Paca St',
  'address2': '',
  'address3': '',
  'city': 'Baltimore',
  'zip_code': '21201',
  'country': 'US',
  'state': 'MD',
  'display_address': ['203 N Paca St', 'Baltimore, MD 21201']},
 'phone': '+14107274898',
 'display_phone': '(410) 727-4898',
 'distance': 1349.560720156645}

In [52]:
# what are the keys for that entry???
json_file['businesses'][0].keys()

dict_keys(['id', 'alias', 'name', 'image_url', 'is_closed', 'url', 'review_count', 'categories', 'rating', 'coordinates', 'transactions', 'price', 'location', 'phone', 'display_phone', 'distance'])

## Identifying "Records"
There is a name for a particular type of json format in which we have a list that contains dictionaries with matching keys for each item... these are called records. 

It looks like "businesses" key is a list of records... **records can be turned into DataFrames for easier inspection**

In [53]:
import pandas as pd
df_businesses = pd.DataFrame(json_file['businesses'])
df_businesses.head()

id                                 alias  \
0  D9A33FM394q99o4QtK5YwA          faidleys-seafood-baltimore-3   
1  u65W69AhbjUlvJJBkEhGNQ        miss-shirleys-cafe-baltimore-9   
2  ieS_5zqxDHcWMCm8BKUYbg  thames-street-oyster-house-baltimore   
3  6am8TZAFnvND52MOz-Yctg     mamas-on-the-half-shell-baltimore   
4  ISn7WyGQaFpsVSRSh0NSqg                sal-and-sons-baltimore   

                         name  \
0            Faidleys Seafood   
1         Miss Shirley's Cafe   
2  Thames Street Oyster House   
3    Mama's On The Half Shell   
4                Sal and Sons   

                                           image_url  is_closed  \
0  https://s3-media3.fl.yelpcdn.com/bphoto/OTjVDC...      False   
1  https://s3-media4.fl.yelpcdn.com/bphoto/9FsOyV...      False   
2  https://s3-media1.fl.yelpcdn.com/bphoto/9hGjo5...      False   
3  https://s3-media2.fl.yelpcdn.com/bphoto/HWY8OF...      False   
4  https://s3-media3.fl.yelpcdn.com/bphoto/LmVL4j...      False   

                                                 url  review_count  \
0  https://www.yelp.com/biz/faidleys-seafood-balt...          1184   
1  https://www.yelp.com/biz/miss-shirleys-cafe-ba...          2919   
2  https://www.yelp.com/biz/thames-street-oyster-...          2729   
3  https://www.yelp.com/biz/mamas-on-the-half-she...          1279   
4  https://www.yelp.com/biz/sal-and-sons-baltimor...           153   

                                          categories  rating  \
0  [{'alias': 'seafood', 'title': 'Seafood'}, {'a...     4.0   
1  [{'alias': 'breakfast_brunch', 'title': 'Break...     4.0   
2  [{'alias': 'seafood', 'title': 'Seafood'}, {'a...     4.5   
3  [{'alias': 'bars', 'title': 'Bars'}, {'alias':...     4.0   
4  [{'alias': 'seafood', 'title': 'Seafood'}, {'a...     4.5   

                                         coordinates        transactions  \
0    {'latitude': 39.291696, 'longitude': -76.62224}          [delivery]   
1  {'latitude': 39.2870995, 'longitude': -76.6053...  [delivery, pickup]   
2     {'latitude': 39.28214, 'longitude': -76.59162}          [delivery]   
3   {'latitude': 39.27986, 'longitude': -76.5752399}  [delivery, pickup]   
4       {'latitude': 39.284, 'longitude': -76.59337}          [delivery]   

  price                                           location         phone  \
0    $$  {'address1': '203 N Paca St', 'address2': '', ...  +14107274898   
1    $$  {'address1': '750 E Pratt St', 'address2': '',...  +14105285373   
2   $$$  {'address1': '1728 Thames St', 'address2': '',...  +14434497726   
3    $$  {'address1': '2901 Odonnell St', 'address2': '...  +14102763160   
4     $  {'address1': '1640 Aliceanna St', 'address2': ...  +14106751466   

    display_phone     distance  
0  (410) 727-4898  1349.560720  
1  (410) 528-5373  1028.736468  
2  (443) 449-7726  2090.712792  
3  (410) 276-3160  3328.825798  
4  (410) 675-1466  1817.406978

In [54]:
# Note this is not perfect... some columns are still off
df_businesses[['categories', 'coordinates', 'location']].head(3)

categories  \
0  [{'alias': 'seafood', 'title': 'Seafood'}, {'a...   
1  [{'alias': 'breakfast_brunch', 'title': 'Break...   
2  [{'alias': 'seafood', 'title': 'Seafood'}, {'a...   

                                         coordinates  \
0    {'latitude': 39.291696, 'longitude': -76.62224}   
1  {'latitude': 39.2870995, 'longitude': -76.6053...   
2     {'latitude': 39.28214, 'longitude': -76.59162}   

                                            location  
0  {'address1': '203 N Paca St', 'address2': '', ...  
1  {'address1': '750 E Pratt St', 'address2': '',...  
2  {'address1': '1728 Thames St', 'address2': '',...

In [55]:
filepath = "D:/Users/Mike PC/Mike PC/.secret/yelp_api.json"

with open(filepath) as f:
    login = json.load(f)
login.keys()

dict_keys(['client-id', 'api-key'])

In [56]:
from yelpapi import YelpAPI
yelp_api = YelpAPI(login['api-key'], timeout_s=5.0)

In [57]:
results = yelp_api.search_query(location='baltimore, MD,21202',
                               term = "Crab Cakes")

type(results)
results.keys()

dict_keys(['businesses', 'total', 'region'])

In [58]:
results['total']

435

In [59]:
import pandas as pd
biz = pd.DataFrame(results['businesses'])
biz.head(2)

id                                 alias  \
0  D9A33FM394q99o4QtK5YwA          faidleys-seafood-baltimore-3   
1  ieS_5zqxDHcWMCm8BKUYbg  thames-street-oyster-house-baltimore   

                         name  \
0            Faidleys Seafood   
1  Thames Street Oyster House   

                                           image_url  is_closed  \
0  https://s3-media3.fl.yelpcdn.com/bphoto/OTjVDC...      False   
1  https://s3-media1.fl.yelpcdn.com/bphoto/9hGjo5...      False   

                                                 url  review_count  \
0  https://www.yelp.com/biz/faidleys-seafood-balt...          1189   
1  https://www.yelp.com/biz/thames-street-oyster-...          2734   

                                          categories  rating  \
0  [{'alias': 'seafood', 'title': 'Seafood'}, {'a...     4.0   
1  [{'alias': 'seafood', 'title': 'Seafood'}, {'a...     4.5   

                                       coordinates transactions price  \
0  {'latitude': 39.291696, 'longitude': -76.62224}   [delivery]    $$   
1   {'latitude': 39.28214, 'longitude': -76.59162}   [delivery]   $$$   

                                            location         phone  \
0  {'address1': '203 N Paca St', 'address2': '', ...  +14107274898   
1  {'address1': '1728 Thames St', 'address2': '',...  +14434497726   

    display_phone     distance  
0  (410) 727-4898  1349.560720  
1  (443) 449-7726  2090.712792

# Pagination vs Offset

You might notice that there are 437 results but only 20 businesses in our list... that is because yelp only shows 20 businesses per 'page'. To get the next 'page' we have to run another call with offset...

In [60]:
results = yelp_api.search_query(location = 'Baltimore, MD,21202',
                               term = 'Crab Cakes',
                               offset = 20)


In [61]:
biz20 = pd.DataFrame(results['businesses'])
biz20.head(2)

id                        alias               name  \
0  n3bY9l_uDccV9X2TxH4rrg  baltimore-seafood-baltimore  Baltimore Seafood   
1  VLGdENTqjFJuCjg1SbW7_g   nicks-fish-house-baltimore   Nicks Fish House   

                                           image_url  is_closed  \
0  https://s3-media3.fl.yelpcdn.com/bphoto/IOF6mr...      False   
1  https://s3-media3.fl.yelpcdn.com/bphoto/xMvMJ9...      False   

                                                 url  review_count  \
0  https://www.yelp.com/biz/baltimore-seafood-bal...           200   
1  https://www.yelp.com/biz/nicks-fish-house-balt...           840   

                                   categories  rating  \
0  [{'alias': 'seafood', 'title': 'Seafood'}]     4.5   
1  [{'alias': 'seafood', 'title': 'Seafood'}]     4.0   

                                      coordinates        transactions price  \
0  {'latitude': 39.28284, 'longitude': -76.58375}  [pickup, delivery]    $$   
1  {'latitude': 39.26036, 'longitude': -76.61421}          [delivery]    $$   

                                            location         phone  \
0  {'address1': '2324 E Boston Sreet', 'address2'...  +14106245166   
1  {'address1': '2600 Insulator Dr', 'address2': ...  +14103474123   

    display_phone     distance  
0  (410) 624-5166  2544.231682  
1  (410) 347-4123  3984.496569

In [62]:
businesses = pd.concat([biz, biz20],
                      ignore_index=True)
display(businesses.head(3), businesses.tail(3))

id                                 alias  \
0  D9A33FM394q99o4QtK5YwA          faidleys-seafood-baltimore-3   
1  ieS_5zqxDHcWMCm8BKUYbg  thames-street-oyster-house-baltimore   
2  u65W69AhbjUlvJJBkEhGNQ        miss-shirleys-cafe-baltimore-9   

                         name  \
0            Faidleys Seafood   
1  Thames Street Oyster House   
2         Miss Shirley's Cafe   

                                           image_url  is_closed  \
0  https://s3-media3.fl.yelpcdn.com/bphoto/OTjVDC...      False   
1  https://s3-media1.fl.yelpcdn.com/bphoto/9hGjo5...      False   
2  https://s3-media4.fl.yelpcdn.com/bphoto/9FsOyV...      False   

                                                 url  review_count  \
0  https://www.yelp.com/biz/faidleys-seafood-balt...          1189   
1  https://www.yelp.com/biz/thames-street-oyster-...          2734   
2  https://www.yelp.com/biz/miss-shirleys-cafe-ba...          2939   

                                          categories  rating  \
0  [{'alias': 'seafood', 'title': 'Seafood'}, {'a...     4.0   
1  [{'alias': 'seafood', 'title': 'Seafood'}, {'a...     4.5   
2  [{'alias': 'breakfast_brunch', 'title': 'Break...     4.0   

                                         coordinates        transactions  \
0    {'latitude': 39.291696, 'longitude': -76.62224}          [delivery]   
1     {'latitude': 39.28214, 'longitude': -76.59162}          [delivery]   
2  {'latitude': 39.2870995, 'longitude': -76.6053...  [delivery, pickup]   

  price                                           location         phone  \
0    $$  {'address1': '203 N Paca St', 'address2': '', ...  +14107274898   
1   $$$  {'address1': '1728 Thames St', 'address2': '',...  +14434497726   
2    $$  {'address1': '750 E Pratt St', 'address2': '',...  +14105285373   

    display_phone     distance  
0  (410) 727-4898  1349.560720  
1  (443) 449-7726  2090.712792  
2  (410) 528-5373  1028.736468

id                      alias             name  \
37  uP4pCpjzUGMDIRzMtHDCkg  old-bay-seafood-baltimore  Old Bay Seafood   
38  pJuoa35eDPRXLf7kUbpMTg     the-choptank-baltimore     The Choptank   
39  6ulR8s_SK3X0wYVQNZ0DSQ  next-phaze-cafe-baltimore  Next Phaze Cafe   

                                            image_url  is_closed  \
37  https://s3-media1.fl.yelpcdn.com/bphoto/ofor5U...      False   
38  https://s3-media3.fl.yelpcdn.com/bphoto/0Mdtr7...      False   
39  https://s3-media3.fl.yelpcdn.com/bphoto/Z9Z5VT...      False   

                                                  url  review_count  \
37  https://www.yelp.com/biz/old-bay-seafood-balti...             5   
38  https://www.yelp.com/biz/the-choptank-baltimor...           200   
39  https://www.yelp.com/biz/next-phaze-cafe-balti...            84   

                                           categories  rating  \
37         [{'alias': 'seafood', 'title': 'Seafood'}]     4.5   
38         [{'alias': 'seafood', 'title': 'Seafood'}]     4.0   
39  [{'alias': 'cafes', 'title': 'Cafes'}, {'alias...     4.0   

                                          coordinates        transactions  \
37  {'latitude': 39.2839579930058, 'longitude': -7...                  []   
38   {'latitude': 39.283338, 'longitude': -76.593321}                  []   
39  {'latitude': 39.2913148, 'longitude': -76.6130...  [pickup, delivery]   

   price                                           location         phone  \
37    $$  {'address1': '333 W Camden St', 'address2': ''...  +18888482473   
38   NaN  {'address1': '1641 Aliceanna St', 'address2': ...  +14437073364   
39    $$  {'address1': '112 E Lexington St', 'address2':...  +14104493232   

     display_phone     distance  
37  (888) 848-2473  1807.250816  
38  (443) 707-3364  1891.591150  
39  (410) 449-3232   710.122481

# Efficient API Practice

In [63]:
# Standard Imports
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
import seaborn as sns

# Additional Imports
import os, json, math, time
from yelpapi import YelpAPI
from tqdm.notebook import tqdm_notebook

In [64]:
# load API creds
with open('../../.secret/yelp_api.json') as f:
    login = json.load(f)
login.keys()

dict_keys(['client-id', 'api-key'])

In [65]:
yelp_api = YelpAPI(login['api-key'], timeout_s=3)
yelp_api

In [66]:
# Define Search Terms
LOCATION = 'Charlottesville, VA'
TERM = 'Winery'

In [67]:
# Create file location to save in progress results
FOLDER = "Data/"

JSON_FILE = f"{FOLDER}results_in_progress_{LOCATION.split(',')[0]}_{TERM}.json"
JSON_FILE

'Data/results_in_progress_Charlottesville_Winery.json'

In [68]:
# Check if the JSON_FILE already exists to prevent overwrite
# IF it doesnt create folders in path and save an empty list

file_exists = os.path.isfile(JSON_FILE)

if file_exists == False:
    
    #Get Folder Name (IF THERE IS ONE)
    folder = os.path.dirname(JSON_FILE)
    
    # Check for folder... if there was a folder in the path make it.
    if len(folder) > 0:
        os.makedirs(folder, exist_ok = True)
    
    print(f"{JSON_FILE} not found. Saving empty list to file.")
    
    with open(JSON_FILE, 'w') as f:
        json.dump([],f)

else:
    print(f"[i] {JSON_FILE} already exists.")

[i] Data/results_in_progress_Charlottesville_Winery.json already exists.


In [69]:
# Determine the number of results already in the file
with open(JSON_FILE, 'r') as f:
    previous_results = json.load(f)

# Set offset based on previous results
n_results = len(previous_results)
print(f'- {n_results} previous results found.')

- 20 previous results found.


In [70]:
results = yelp_api.search_query(location=LOCATION,
                               term=TERM,
                               offset = n_results+1)
results.keys()

dict_keys(['businesses', 'total', 'region'])

In [71]:
# Determine the total number of businesses captured by the Yelp Search
total_results = results['total']
total_results

77

In [72]:
# Determine the number saved per page
results_per_page = len(results['businesses'])
results_per_page

20

In [73]:
# Use math.ceil to round up for the toal number of pages needed
import math, time

n_pages = math.ceil((total_results-n_results) / results_per_page)
n_pages

# Subtract by n_results to avoid duplicating previous work!!!

3

In [74]:
# Save our results as we loop...
previous_results.extend(results['businesses'])
with open(JSON_FILE, 'w') as f:
    json.dump(previous_results,f)

In [75]:
# Look what tqdm can do!
from tqdm.notebook import tqdm_notebook

for i in tqdm_notebook(range(n_pages)):
    time.sleep(.5)

  0%|          | 0/3 [00:00<?, ?it/s]

In [76]:
for i in tqdm_notebook(range(1,n_pages+1)):
    time.sleep(.2)
    
    # Read in existing results for API call
    with open(JSON_FILE, 'r') as f:
        previous_results = json.load(f)
    
    # check length of results (IE are there results?)
    n_results = len(previous_results)
    
    # Run API using n_results+1 as offset
    results = yelp_api.search_query(location = LOCATION,
                                   term = TERM,
                                   offset = n_results+1)
    
    previous_results.extend(results['businesses'])
    
    with open(JSON_FILE, 'w') as f:
        json.dump(previous_results, f)

  0%|          | 0/3 [00:00<?, ?it/s]

In [80]:
final_df = pd.read_json(JSON_FILE)
display(final_df.head(), final_df.tail())

id                                        alias  \
0  9WSh050KKnLktbzE3X5KBA          jefferson-vineyards-charlottesville   
1  s6pzoTZDjQ4h_SY_FpP2SQ              gabriele-rausse-charlottesville   
2  Z9djBo-FcFChFjE7sfXaIw  pippin-hill-farm-and-vineyards-north-garden   
3  8H_FKdb3PRx1RLoAjnE86w        merrie-mill-farm-and-vineyard-keswick   
4  gXNjZS7yl2MRq9BY63v8Tg                glass-house-winery-free-union   

                           name  \
0           Jefferson Vineyards   
1               Gabriele Rausse   
2  Pippin Hill Farm & Vineyards   
3   Merrie Mill Farm & Vineyard   
4            Glass House Winery   

                                           image_url  is_closed  \
0  https://s3-media4.fl.yelpcdn.com/bphoto/MGBDPm...      False   
1  https://s3-media1.fl.yelpcdn.com/bphoto/HS8EsX...      False   
2  https://s3-media2.fl.yelpcdn.com/bphoto/lc2NOq...      False   
3  https://s3-media2.fl.yelpcdn.com/bphoto/SOP99o...      False   
4  https://s3-media2.fl.yelpcdn.com/bphoto/MHRI2-...      False   

                                                 url  review_count  \
0  https://www.yelp.com/biz/jefferson-vineyards-c...           237   
1  https://www.yelp.com/biz/gabriele-rausse-charl...            77   
2  https://www.yelp.com/biz/pippin-hill-farm-and-...           326   
3  https://www.yelp.com/biz/merrie-mill-farm-and-...            26   
4  https://www.yelp.com/biz/glass-house-winery-fr...            87   

                                          categories  rating  \
0       [{'alias': 'wineries', 'title': 'Wineries'}]     4.0   
1       [{'alias': 'wineries', 'title': 'Wineries'}]     4.5   
2  [{'alias': 'venues', 'title': 'Venues & Event ...     4.0   
3  [{'alias': 'winetastingroom', 'title': 'Wine T...     4.5   
4  [{'alias': 'wineries', 'title': 'Wineries'}, {...     4.5   

                                         coordinates        transactions  \
0   {'latitude': 37.990044, 'longitude': -78.443643}                  []   
1  {'latitude': 37.9460038, 'longitude': -78.4832...                  []   
2  {'latitude': 37.9577015, 'longitude': -78.661157}  [pickup, delivery]   
3  {'latitude': 38.0542984, 'longitude': -78.2956...                  []   
4  {'latitude': 38.2025939, 'longitude': -78.5791...                  []   

  price                                           location         phone  \
0    $$  {'address1': '1353 Thomas Jefferson Pkwy', 'ad...  +14349773042   
1    $$  {'address1': '3247 Carters Mountain Rd', 'addr...  +14349811677   
2    $$  {'address1': '5022 Plank Rd', 'address2': '', ...  +14342028063   
3   NaN  {'address1': '594 Merrie Mill Farm', 'address2...  +14343653006   
4    $$  {'address1': '5898 Free Union Rd', 'address2':...  +14349750094   

    display_phone      distance  
0  (434) 977-3042   6630.950399  
1  (434) 981-1677  10995.036917  
2  (434) 202-8063  18452.142160  
3  (434) 365-3006  16223.671596  
4  (434) 975-0094  19646.202482

id                                           alias  \
40  5qNkM1pPktyaQVbQfjZfdw        top-shelf-transportation-charlottesville   
41  RWy9KbL58h9HhTfJOHhmJA                     vingo-tours-charlottesville   
42  maMeVkLA1yIBkzbOm7aF6A     hyatt-place-charlottesville-charlottesville   
43  OePDEUHpth8EkGAbfQ7FGg      omni-charlottesville-hotel-charlottesville   
44  yQ27NzO_Odi1gsjdv-XpfQ  charlottesville-transportation-charlottesville   

                              name  \
40        Top Shelf Transportation   
41                     VinGo Tours   
42     Hyatt Place Charlottesville   
43      Omni Charlottesville Hotel   
44  Charlottesville Transportation   

                                            image_url  is_closed  \
40  https://s3-media1.fl.yelpcdn.com/bphoto/OQZ4jE...      False   
41  https://s3-media1.fl.yelpcdn.com/bphoto/AQUza5...      False   
42  https://s3-media2.fl.yelpcdn.com/bphoto/5uIHa5...      False   
43  https://s3-media4.fl.yelpcdn.com/bphoto/yRMk-B...      False   
44  https://s3-media3.fl.yelpcdn.com/bphoto/Di9WTU...      False   

                                                  url  review_count  \
40  https://www.yelp.com/biz/top-shelf-transportat...            22   
41  https://www.yelp.com/biz/vingo-tours-charlotte...             2   
42  https://www.yelp.com/biz/hyatt-place-charlotte...            50   
43  https://www.yelp.com/biz/omni-charlottesville-...           109   
44  https://www.yelp.com/biz/charlottesville-trans...            13   

                                           categories  rating  \
40  [{'alias': 'winetours', 'title': 'Wine Tours'}...     5.0   
41    [{'alias': 'winetours', 'title': 'Wine Tours'}]     5.0   
42  [{'alias': 'hotels', 'title': 'Hotels'}, {'ali...     4.0   
43  [{'alias': 'hotels', 'title': 'Hotels'}, {'ali...     3.5   
44  [{'alias': 'airport_shuttles', 'title': 'Airpo...     5.0   

                                          coordinates transactions price  \
40   {'latitude': 38.086855, 'longitude': -78.469888}           []   NaN   
41  {'latitude': 38.0423698425293, 'longitude': -7...           []   NaN   
42  {'latitude': 38.0664497284703, 'longitude': -7...           []    $$   
43   {'latitude': 38.031274, 'longitude': -78.483499}           []   $$$   
44  {'latitude': 38.0423698425293, 'longitude': -7...           []   NaN   

                                             location         phone  \
40  {'address1': '1869 Seminole Trl', 'address2': ...  +14342959462   
41  {'address1': None, 'address2': None, 'address3...  +14342601603   
42  {'address1': '2100 Bond St', 'address2': '', '...  +14349955200   
43  {'address1': '212 Ridge McIntire Rd', 'address...  +14349715500   
44  {'address1': None, 'address2': None, 'address3...  +14344668465   

     display_phone     distance  
40  (434) 295-9462  4728.832738  
41  (434) 260-1603  1727.472447  
42  (434) 995-5200  2763.470629  
43  (434) 971-5500  1577.078252  
44  (434) 466-8465  1727.472447

In [82]:
final_df.duplicated(subset = 'id').sum()

0

In [36]:
type(previous_results)

list

In [83]:
# Save final DF to a csv (or csv.gz if you want to compress)
final_df.to_csv(f"{FOLDER}results_final_{LOCATION.split(',')[0]}_{TERM}.csv.gz",
               compression='gzip', index = False)